test/unittests/execute_call_test.cpp

```diff
@@ -622,6 +643,100 @@ TEST(execute_call, call_max_depth)
     EXPECT_THAT(execute(*instance, 1, {}, MaxDepth), Traps());
 }
 
+TEST(execute_call, execute_imported_max_depth)
+{
+    /* wat2wasm
+    (import "mod" "foo" (func))
+    (func)
+    */
+
+    const auto wasm =
+        from_hex("0061736d01000000010401600000020b01036d6f6403666f6f0000030201000a040102000b");
+
+    auto module = parse(wasm);
+    auto host_foo = [](Instance& /*instance*/, const Value*, int depth) -> ExecutionResult {
+        EXPECT_LE(depth, MaxDepth);
+        return Void;
+    };
+    const auto host_foo_type = module->typesec[0];
+
+    auto instance = instantiate(std::move(module), {{host_foo, host_foo_type}});
+
+    EXPECT_THAT(execute(*instance, 0, {}, MaxDepth), Result());
+    EXPECT_THAT(execute(*instance, 1, {}, MaxDepth), Result());
+    EXPECT_THAT(execute(*instance, 0, {}, MaxDepth + 1), Traps());
+    EXPECT_THAT(execute(*instance, 1, {}, MaxDepth + 1), Traps());
+}
+
+TEST(execute_call, imported_function_from_another_module_max_depth)
+{
+    /* wat2wasm
+    (module
+      (func)
+      (export "f" (func 0))
+    )
+    */
+    const auto bin1 = from_hex("0061736d0100000001040160000003020100070501016600000a040102000b");
+    auto module1 = parse(bin1);
+    auto instance1 = instantiate(std::move(module1));
+
+    /* wat2wasm
+    (module
+      (func (import "m1" "f"))
+      (func)
+      (func call 0)
+      (func call 1)
+    )
+    */
+    const auto bin2 = from_hex(
+        "0061736d01000000010401600000020801026d31016600000304030000000a0e0302000b040010000b04001001"
+        "0b");
+    auto module2 = parse(bin2);
+
+    const auto func_idx = fizzy::find_exported_function(*instance1->module, "f");
+    ASSERT_TRUE(func_idx.has_value());
+
+    auto sub = [&instance1, func_idx](Instance&, const Value* args, int depth) -> ExecutionResult {
+        return fizzy::execute(*instance1, *func_idx, args, depth + 1);
+    };
+
+    auto instance2 = instantiate(std::move(module2), {{sub, instance1->module->typesec[0]}});
+
+    EXPECT_THAT(execute(*instance2, 2, {}, MaxDepth - 1), Traps());
+    EXPECT_THAT(execute(*instance2, 3, {}, MaxDepth - 1), Result());
+}
+
+TEST(execute_call, count_calls_to_imported_function)
+{
+    /* wat2wasm
+    (global $counter (import "m0" "counter") (mut i64))
+    (func $infinite (export "infinite")
+      (global.set $counter (i64.add (global.get $counter) (i64.const 1)))
+      (call $infinite)
+    )
+    */
+    const auto wasm1 = from_hex(
+        "0061736d01000000010401600000020f01026d3007636f756e746572037e0103020100070c0108696e66696e69"
+        "746500000a0d010b00230042017c240010000b");
+
+    Value counter;
+
+    auto i1 = instantiate(parse(wasm1), {}, {}, {}, {{&counter, {ValType::i64, true}}});
+    counter.i64 = 0;
+    EXPECT_THAT(execute(*i1, 0, {}), Traps());
+    EXPECT_EQ(counter.i64, MaxDepth + 1);  // FIXME: Should be MaxDepth.
```

Comment:
Why should it be? Function is executed `MaxDepth + 1` times, because depth starts with 0 and ends with `MaxDepth` inclusive.

Reply:
Oh, this is correct, but I prefer there to be 2048 allowed calls, not 2049. This is changed in #669.

Reply:
I removed the FIXME comment as we have #669 to discuss it.